### ISSUE TITLE
detect unused static functions in local classes/structs

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#72070]
```cpp
class C
{
    static void s_f() {}
};

struct S
{
    static void s_f() {}
};

namespace {
    void anon_f() {}
}

namespace ns {
    static void ns_f() {}
}

static void f() {}
```

```
<source>:12:10: warning: unused function 'anon_f' [-Wunused-function]
   12 |     void anon_f() {}
      |          ^~~~~~
<source>:16:17: warning: unused function 'ns_f' [-Wunused-function]
   16 |     static void ns_f() {}
      |                 ^~~~
<source>:19:13: warning: unused function 'f' [-Wunused-function]
   19 | static void f() {}
      |             ^
```

https://godbolt.org/z/9s9YG5G7j

One could argue that an unused regular method in a local class/struct (not in the example above) is similar to this and could also be diagnosed.

**Comment by Lancern** [comment]
Clang do emit warnings on unused methods in local classes and structs:

```cpp
void test();

void test() {
  struct Local {
    void foo() {}
    void bar() {}
  };

  Local l;
  l.foo();
}
```

```
<source>:6:10: warning: unused member function 'bar' [-Wunused-member-function]
    6 |     void bar() {}
      |          ^~~
1 warning generated.
```

See https://godbolt.org/z/fP6vxWjcM.

**Comment by firewave** [comment]
> Clang do emit warnings on unused methods in local classes and structs:

They are also emitted when the struct is in an anonymous namespace. And it also applies to static methods.

https://godbolt.org/z/Wbax3zx34

**Comment by Lancern** [comment]
> > Clang do emit warnings on unused methods in local classes and structs:
> 
> They are also emitted when the struct is in an anonymous namespace. And it also applies to static methods.
> 
> https://godbolt.org/z/Wbax3zx34

Unused static member functions in local classes are also diagnosed:

```cpp
void test();

void test() {
  struct Local {
    static void foo() {}
    static void bar() {}
  };

  Local l;
  l.foo();
```

```
<source>:6:17: warning: unused member function 'bar' [-Wunused-member-function]
    6 |     static void bar() {}
      |                 ^~~
1 warning generated.
```

So what is the exact problem? From the examples, it seems that methods in local classes do not have any differences on the diagnostics you mentioned. Unused methods in local classes are diagnosed. Could you please provide an example that shows the exact problem you spotted? An example that clang does not generate diagnostics but it should do.

**Comment by firewave** [comment]
> Could you please provide an example that shows the exact problem you spotted?

My original example shows that no warnings are produced for `C::s_f()` and `S::s_f()`.

**Comment by Lancern** [comment]
> > Could you please provide an example that shows the exact problem you spotted?
> 
> My original example shows that no warnings are produced for `C::s_f()` and `S::s_f()`.

OK. This looks reasonable. I would like to take this issue.

**Comment by Lancern** [comment]
The static methods `C::s_f` and `S::s_f` in the example is different from other functions. Clang emits warnings for `anon_f`, `ns_f`, and `f` because they have _internal linkage_. However, static member functions `C::s_f` and `S::s_f` have _external linkage_. So they are different stories.

To be accurate, the diagnostic improvements proposed in this issue are actually "unused inline functions in a TU should be diagnosed". Is this really what we want?

**Comment by Lancern** [comment]
> We can't mark member functions (static or non-static) of non-local types as unused, because they might be used in some other translation unit.

Maybe an unused inline function is a potential point for diagnostics? Here `C::s_f` and `S::s_f` are such examples.

Of course, only those inline functions defined in the primary source file are diagnostic targets. Those inline functions introduced by header files are not to be diagnosed.

**Comment by firewave** [comment]
> The static methods `C::s_f` and `S::s_f` in the example is different from other functions. Clang emits warnings for `anon_f`, `ns_f`, and `f` because they have _internal linkage_. However, static member functions `C::s_f` and `S::s_f` have _external linkage_. So they are different stories.

I knew I missed something about this. Any documentation/specification you could point me to so I could read up on this?

But what if you are not "exporting" these (default in a Windows) or have `hidden` visibility set?

**Comment by dwblaikie** [comment]
not exporting doesn't mean you are only using them in a single file, I tihnk? exporting is about using across DLL boundaries, right? So we still can't be sure you aren't using them in another .cpp file within the same DLL?

> Of course, only those inline functions defined in the primary source file are diagnostic targets. Those inline functions introduced by header files should not be diagnosed.

I don't think we generally do that (have diagnostics that only trigger on the primary source file) - might be of limited value (folks should be putting classes/things in the primary source file in an anonymous namespace, etc - which clang-tidy might be more suited to)

**Comment by firewave** [comment]
> not exporting doesn't mean you are only using them in a single file, I tihnk? exporting is about using across DLL boundaries, right? So we still can't be sure you aren't using them in another .cpp file within the same DLL?

Yes. And visibility is only about shared objects. So I think this can be closed.

These are all points that I were already aware off *before* I filed the ticket but somehow my mind blanked. Sorry about that.

**Comment by dwblaikie** [comment]
> These are all points that I were already aware off before I filed the ticket but somehow my mind blanked. Sorry about that.

No worries - thanks for talking it through.
